doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,969 @@
+.. _ltac2:
+
+Ltac2
+=====
+
+The Ltac tactic language is probably one of the ingredients of the success of
+Coq, yet it is at the same time its Achilles' heel. Indeed, Ltac:
+
+- has often unclear semantics
+- is very non-uniform due to organic growth
+- lacks expressivity (data structures, combinators, types, ...)
+- is slow
+- is error-prone and fragile
+- has an intricate implementation
+
+Following the need of users that start developing huge projects relying
+critically on Ltac, we believe that we should offer a proper modern language
+that features at least the following:
+
+- at least informal, predictable semantics
+- a typing system
+- standard programming facilities (i.e. datatypes)
+
+This new language, called Ltac2, is described in this chapter. It is still
+experimental but we encourage nonetheless users to start testing it,
+especially wherever an advanced tactic language is needed. The previous
+implementation of Ltac, described in the previous chapter, will be referred to
```

Comment:
Experimental means that the APIs are prone to change without previous notice. I don't know whether α encompasses this kind of feature.

> We should have a much more positive introduction to the chapter to motivate users to begin using the feature

I am quite confident that Ltac1 is a motivation strong enough to switch to Ltac2, but I am obviously biased...

Reply:
Then we should say in the text "the APIs are subject to change without notice" but also state that the goal is to eventually retire Ltac1.  And presumably API changes would be primarily for bugs, conversion problems and design issues.

Maybe also say that we'd like to work with users who are willing to try to convert existing Ltac1 scripts--meaning someone would make an effort to be responsive to questions/issues and make fixes on a timely basis (perhaps quicker than the next official release; maybe a branch).  If that's something you or others are willing to undertake.  Otherwise I can imagine a bunch of users running into the same issues repeatedly, a waste of time for both users and developers.  Also less likely for the new code to possibly get a bad reputation due to early issues.  Also, without this, the transition time to get the community onto Ltac2 could be many years.

> I am quite confident that Ltac1 is a motivation strong enough to switch to Ltac2, but I am obviously biased...

How about new users?  The Ltac1 documentation, poor as it is, is more complete than the Ltac2 doc.  I'm not sure what choice I would make.  (How about people who gave up on Ltac1?)

Also, are there important proof libraries that use Ltac1 in a way that prevents using them with Ltac2?  If so, the incompatibility would block their users from using Ltac2.  It might be worth some time to identify these libraries and their interdependencies.  (That might identify good candidates for the trial conversions I suggested above.)  Also, I recall seeing info on calling Ltac1 from Ltac2 but I don't remember seeing info on calling Ltac2 from Ltac1.  Are the restrictions documented?

Of course, this all depends on what people are willing to spend their time on.

Reply:
I am postponing the change of this paragraph after the merge. This kind of introductory text has to be written before the next release, and needs a bit of serious pondering.

Reply:
Sounds good.  If you wish, I'd be happy to review the change and/or help with wording.